crates/viewer/re_space_view_dataframe/src/display_record_batch.rs

```diff
@@ -0,0 +1,332 @@
+//! Intermediate data structures to make `re_datastore`'s schemas and [`RecordBatch`]s more amenable
+//! to displaying in a table.
+
+use thiserror::Error;
+
+use re_chunk_store::external::re_chunk::external::arrow2::{
+    array::{
+        Array as ArrowArray, DictionaryArray as ArrowDictionaryArray, ListArray as ArrowListArray,
+        PrimitiveArray as ArrowPrimitiveArray, StructArray as ArrowStructArray,
+    },
+    datatypes::DataType,
+    datatypes::DataType as ArrowDataType,
+};
+use re_chunk_store::{ColumnDescriptor, ComponentColumnDescriptor, LatestAtQuery, RowId};
+use re_dataframe::RecordBatch;
+use re_log_types::{EntityPath, TimeInt, Timeline};
+use re_types::external::arrow2::datatypes::IntegerType;
+use re_types_core::ComponentName;
+use re_ui::UiExt;
+use re_viewer_context::{UiLayout, ViewerContext};
+
+#[derive(Error, Debug)]
+pub(crate) enum DisplayRecordBatchError {
+    #[error("Unknown control column: {0}")]
+    UnknownControlColumn(String),
+
+    #[error("Unexpected column data type for timeline '{0}': {1:?}")]
+    UnexpectedTimeColumnDataType(String, ArrowDataType),
+
+    #[error("Unexpected column data type for component '{0}': {1:?}")]
+    UnexpectedComponentColumnDataType(String, ArrowDataType),
+}
+
+pub(crate) enum ComponentData {
+    Null,
+    ListArray(ArrowListArray<i32>),
+    DictionaryArray {
+        dict: ArrowDictionaryArray<i32>,
+        values: ArrowListArray<i32>,
+    },
+}
+
+impl ComponentData {
+    #[allow(clippy::borrowed_box)] // https://github.com/rust-lang/rust-clippy/issues/11940
+    fn try_new(
+        descriptor: &ComponentColumnDescriptor,
+        column_data: &Box<dyn ArrowArray>,
+    ) -> Result<Self, DisplayRecordBatchError> {
+        match column_data.data_type() {
+            DataType::Null => Ok(Self::Null),
+            DataType::List(_) => Ok(Self::ListArray(
+                column_data
+                    .as_any()
+                    .downcast_ref::<ArrowListArray<i32>>()
+                    .expect("sanity checked")
```

Comment:
If this error can't happen, we should use `.unwrap()` and explain why it can't happen.

If it _can_ happen, we should return a proper `Err` instead.

Crashing with "sanity checked" as the message is not very helpful to anyone

Reply:
I'm surprised by this comment. It was my understanding that `expect(reason)` _precisely_ means "unwrap-that-cannot-happen-because-$reason".

In this particular case, it cannot happen (short of a bug in `re_dataframe`). This is pattern that's often used in Clement's code.

Reply:
Fair enough, but then the error message should be something useful:

`.expect("There is bug in the code at …")`

Reply:
OK, I've updated the message.